### PR TITLE
DAOS-7821 cq: Add a commit-hook for running pylint.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a --diff option to daos_pylint and run it from a commit hook.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: utils/cq/daos_pylint.py [hunk 1]
```diff
@@ -496,8 +527,38 @@ def main():
...
+                file = parts[3][2:-1]
+                regions = OutPutRegion()
+                all_files.add_regions(file, regions)
+                if not args.git:
+                    all_files.add(file)
+                continue
+            if line.startswith('@@ '):
+                parts = line.split(' ')
+                (post_start, post_len) = parts[2][1:].split(',')
+                regions.add_region(int(post_start), int(post_len))
```
Comment (ashleypittman) [inline utils/cq/daos_pylint.py:551]: There might be an off-by-one error here although given that we're using diff -U10 it probably doesn't matter in practice.

File: .github/workflows/pylint.yml [hunk 2]
```diff
@@ -20,5 +20,11 @@ jobs:
         run: python3 -m pip install --requirement utils/cq/requirements.txt
       - name: Install enchant
         run: sudo apt-get update && sudo apt-get -y install enchant
+      - name: Pull via git
+        run: git fetch origin ${{ github.event.pull_request.base.sha }}
+      - name: See git diff
+        run: git diff ${{ github.event.pull_request.base.sha }} -U10
```
Comment (ashleypittman) [inline .github/workflows/pylint.yml:26]: This is debugging only so could come out but it's also not doing any harm for now.

**Review by jolivier23 — APPROVED** [review]

File: utils/cq/daos_pylint.py [hunk 3]
```diff
@@ -235,189 +240,201 @@ def run(self, args):
...
             for file in self.scons_files:
-                if parse_file(args, file, scons=True):
+                if self.parse_file(args, file, scons=True):
                     failed = True
         return failed
 
+    def parse_file(self, args, target_file, ftest=False, scons=False, fake_scons=False):
+        """Parse a list of targets.
+
+        Returns True if warnings issued to GitHub."""
```
Comment (daltonbohning) [inline utils/cq/daos_pylint.py:260]: FYI - In ftest at least, we follow this format
```suggestion
        """Parse a list of targets.

        Args:
            args (<type>): <desc>
            ...

        Returns:
            bool: True if warnings issued to GitHub.
        """
```

File: utils/cq/daos_pylint.py [hunk 4]
```diff
@@ -235,189 +240,201 @@ def run(self, args):
...
+
+        # pylint: disable=too-many-branches
+
+        def word_is_allowed(word, code):
+            """Return True is misspelling is permitted"""
```
Comment (daltonbohning) [inline utils/cq/daos_pylint.py:265]:
```suggestion
            """Return True if misspelling is permitted"""
```

File: utils/cq/daos_pylint.py [hunk 5]
```diff
@@ -235,189 +240,201 @@ def run(self, args):
...
+
+            # pylint: disable=too-many-return-statements
+
+            # Skip short words for now to cut down on noise whilst we resolve existing issues.
+            if len(word) < 5:
+                return True
```
Comment (daltonbohning) [inline utils/cq/daos_pylint.py:271]: FYI that this structure works (example only)
```
checks = [
     lambda: word in code,
     lambda: word == code,
     lambda: code.endswith(word)
]
return any(c() for c in checks) # <- lazy evaluation
```

**Review by daltonbohning — APPROVED** [review]
I didn't run this locally, but conceptually, LGTM

**Pull request merged**